File: Visualizer/src/main/java/com/github/ezauton/visualizer/processor/RecordingDataProcessor.java
package com.github.ezauton.visualizer.processor;

import com.github.ezauton.recorder.Recording;
import com.github.ezauton.visualizer.processor.factory.DataProcessorFactory;
import com.github.ezauton.visualizer.util.DataProcessor;
import com.github.ezauton.visualizer.util.Environment;
import javafx.animation.Interpolator;
import javafx.animation.KeyValue;

import java.util.*;

public class RecordingDataProcessor implements DataProcessor {

    final List<DataProcessor> childDataProcessors = new ArrayList<>();

    public RecordingDataProcessor(Recording recording, DataProcessorFactory factory) {
        recording.getRecordingMap()
                .values()
                .stream()
                .map(factory::getProcessor)
                .filter(Optional::isPresent)
                .map(Optional::get)
                .forEach(childDataProcessors::add);
    }

    @Override
    public void initEnvironment(Environment environment) {
        for (DataProcessor d : childDataProcessors) {
            if (d != null) {
                d.initEnvironment(environment);
            }
        }
    }

    @Override
    public Map<Double, List<KeyValue>> generateKeyValues(Interpolator interpolator) {
        Map<Double, List<KeyValue>> ret = new HashMap<>();
        for (DataProcessor dataProcessor : childDataProcessors) {
            if (dataProcessor != null) {
                Map<Double, List<KeyValue>> keyValMap = dataProcessor.generateKeyValues(interpolator);
                if (keyValMap != null) {
                    for (Map.Entry<Double, List<KeyValue>> entry : keyValMap.entrySet()) {
                        if (!ret.containsKey(entry.getKey())) // not contained
                        {
                            ret.put(entry.getKey(), new ArrayList<>(entry.getValue()));
                        } else { // contained
                            ret.get(entry.getKey()).addAll(entry.getValue());
                        }
                    }
                }
            }
        }
        return ret;
    }
}

File: Visualizer/src/main/java/com/github/ezauton/visualizer/processor/RecordingDataProcessor.kt
package com.github.ezauton.visualizer.processor

import com.github.ezauton.recorder.Recording
import com.github.ezauton.visualizer.processor.factory.DataProcessorFactory
import com.github.ezauton.visualizer.util.DataProcessor
import com.github.ezauton.visualizer.util.Environment
import javafx.animation.Interpolator
import javafx.animation.KeyValue

class RecordingDataProcessor(recording: Recording, factory: DataProcessorFactory) : DataProcessor {

  private val childDataProcessors = recording.recordingMap.values
    .asSequence()
    .map { subRecording -> factory.getProcessor(subRecording) }
    .filterNotNull()
    .toList()


  override fun initEnvironment(environment: Environment) {
    childDataProcessors.forEach {
      it.initEnvironment(environment)
    }
  }

  override fun generateKeyValues(interpolator: Interpolator): Map<Double, List<KeyValue>> {
    val ret: MutableMap<Double, MutableList<KeyValue>> = HashMap()
    for (dataProcessor in childDataProcessors) {
      val keyValMap = dataProcessor.generateKeyValues(interpolator)
      for ((key, value) in keyValMap) {
        if (!ret.containsKey(key)) // not contained
        {
          ret[key] = ArrayList(value)
        } else { // contained
          ret[key]!!.addAll(value)
        }
      }
    }
    return ret
  }

}
